[distributed][nccl] Log NCCL version in NCCL errors.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26583 [distributed][nccl] Log NCCL version in NCCL errors.**

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

rohan-varma committed:
[wip][pytorch][distributed] add function to get NCCL version for logging

Adds a function that uses the nccl api to get the version code. Converts it to a readable version. Will be
used for logging NCCL version in exception messages.

Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -16,11 +16,25 @@
     }                                                                     \
   } while (0)
 
+inline std::string getNcclVersion() {
```
I'm not sure what's the best way to define this function `NCCLUtils.hpp`. If I put it in an anonymous namespace, then it won't be available to `ProcessGroupNCCL`, but I need it there. If I don't inline the function the linker gives multiple definition errors. Any suggestions?

pietern replied:
Without its own translation unit you either have to inline or put it in an anonymous namespace in this header (which is essentially equivalent to inline). You can avoid duplication by adding a `NCCLUtils.cpp` file and adding it in there. If you do so, you can also make this use a static `once_flag` and ensure you only do this once, in case this routine is abused and called a lot some time in the future.

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses the nccl api to get the version code. Converts it to a readable version. Will be
used for logging NCCL version in exception messages.
Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses the nccl api to get the version code. Converts it to a readable version. Will be
used for logging NCCL version in exception messages.
Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -16,11 +16,25 @@
           std::to_string(__LINE__) + ", " +                               \
-          std::string(ncclGetErrorString(error));                         \
+          std::string(ncclGetErrorString(error)) + ", NCCL version " +    \
+          getNcclVersion();                                               \
```
You can turn this into a custom `ncclGetErrorString` that tacks on the version.

Then the error below will automatically use the same as well.

rohan-varma replied:
Would it be confusing to override `ncclGetError` since it's implemented in the NCCL API? What about a new function `ncclGetErrorWithVersion` that wraps `ncclGetErrorString` and tacks on the version info?

pietern replied:
Yep, that sounds good.

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

pietern commented:
@pytorchbot retest this please

pietern reviewed (approved):
Thanks, @rohan-varma. LGTM.

pietern commented on torch/lib/c10d/CMakeLists.txt:
```diff
@@ -53,6 +53,7 @@ set(C10D_LIBS torch)
 
 if(USE_C10D_NCCL)
   list(APPEND C10D_SRCS ProcessGroupNCCL.cpp)
```
You can add multiple sources on a single line, FWIW.

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma commented:
@pietern - sorry, could you take another look? I just updated the PR to use the `once_flag` as you suggested. The flag is used along with `std::call_once` to ensure the nccl API is only directly called once, basically caching the result. Hopefully it's in line with what you were thinking.

pietern commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -0,0 +1,32 @@
+std::once_flag ncclGetVersionFlag;
+
+std::string getNcclVersion() {
+  std::string versionString;
```
This needs to be static or you'll return an empty string on the second call.

You can also move the `once_flag` into this function as a static to isolate it.

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][nccl] Log NCCL version in NCCL errors."


Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```
Differential Revision: [D17473200](https://our.internmc.facebook.com/intern/diff/D17473200/)

[ghstack-poisoned]

pietern commented:
This was reverted in 4bd1da14588a417e0b50c48e3b03726ea991ede6. Can you resubmit?
